returnn/tf/util/data.py

```diff
@@ -2337,6 +2337,32 @@ def copy_extend_batch(self, batch):
         data.placeholder = x
     return data
 
+  def copy_unpack_batch(self):
+    """
+    If the batch dim contains a :class:`BatchInfo.PackedDim`, unpack it and restore the data from before the packing.
+
+    :rtype: Data
+    """
+    assert self.have_batch_axis()
+    assert self.batch, "%s: batch unset" % self
+    data = self.copy()
+    kwargs = self.get_kwargs(include_special_axes=False)
+
+    dim_tags = []
+    for dim_tag in data.dim_tags:
+      if dim_tag.is_batch_dim() and hasattr(dim_tag.batch, "virtual_dims"):
+        for virtual_dim in dim_tag.batch.virtual_dims:
+          if isinstance(virtual_dim, BatchInfo.PackedDim):
+            dim_tags.append(virtual_dim.dim_tag)
+            kwargs["batch"] = virtual_dim.dim_tag.batch
+          else:
+            dim_tags.append(DimensionTag(kind=DimensionTag.Types.Batch))
```

Comment:
This doesn't really make sense.
You should create a new function in `BatchInfo` similar to `copy_remove_beam`, like `copy_remove_dim` or so.
And then the code above should be like:
```suggestion
        batch = dim_tag.batch
        new_batch_dim_tag = None
        for virtual_dim in batch.virtual_dims:
          if isinstance(virtual_dim, BatchInfo.PackedDim):
            dim_tags.append(virtual_dim.dim_tag)
            batch = batch.copy_remove_dim(virtual_dim)
          elif not new_batch_dim_tag:
            new_batch_dim_tag = DimensionTag(kind=DimensionTag.Types.Batch)
            dim_tags.append(new_batch_dim_tag)
        assert new_batch_dim_tag
        new_batch_dim_tag.batch = batch
        kwargs["batch"] = batch
```